% Created 2013-08-09 Fri 12:27
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{fixltx2e}
\usepackage{graphicx}
\usepackage{longtable}
\usepackage{float}
\usepackage{wrapfig}
\usepackage{soul}
\usepackage{textcomp}
\usepackage{marvosym}
\usepackage{wasysym}
\usepackage{latexsym}
\usepackage{amssymb}
\usepackage{amstext}
\usepackage{hyperref}
\tolerance=1000
\usepackage{fullpage}
\author{Chris Wallace}
\date{2013-05-22 Wed}
\title{annotSnpStats: a package for annotated SnpMatrix objects}
\hypersetup{
  pdfkeywords={},
  pdfsubject={},
  pdfcreator={Emacs 24.2.1 (Org mode 8.0.3)}}
\begin{document}

\maketitle
\tableofcontents

%\VignetteIndexEntry{Colocalisation analysis}

\section{Introduction}
\label{sec-1}

\texttt{annotSnpStats} defines the \texttt{aSnpMatrix} class which extends the
\texttt{SnpMatrix} class from \texttt{snpStats} by aligning two dataframes contain
SNP (column) specific and sample (row) specific information.
Subsetting, binding etc can then be carried out keeping these
additional information aligned.
\section{Example}
\label{sec-2}

Use some example data from the \texttt{snpStats} package to define some new
\texttt{aSnpMatrix} objects and play with them.

\subsection{Create new objects}
\label{sec-2-1}
<<>>=
library(snpStats)
library(annotSnpStats)
data(testdata, package="snpStats")
i1 <- 1:20
j1 <- 1:100

## initialize
rownames(Asnps) <- colnames(Autosomes)
rownames(Xsnps) <- colnames(Xchromosome)

as11 <- new("aSnpMatrix",
                 .Data=Autosomes[i1,j1],
                 snps=Asnps[j1,,drop=FALSE],
                 samples=subject.data[i1,])
as11
@ %def
\subsection{Access annotation objects}
\label{sec-2-2}

<<>>=
head(as11@snps)
head(as11@samples)
@ %def
\subsection{Use in place of regular SnpMatrix objects}
\label{sec-2-3}

\texttt{snpStats} functions should all work with \texttt{aSnpMatrix} objects

<<>>=
summary(as11)
@ %def
\subsection{Subsetting}
\label{sec-2-4}

Subsetting subsets both the \texttt{SnpMatrix} and the annotation objects

<<>>=
asub <- as11[1:3,1:5]
dim(as11)
dim(asub)
nrow(as11@snps)
nrow(asub@snps)
nrow(as11@samples)
nrow(asub@samples)
@ %def
\subsection{Binding}
\label{sec-2-5}
You can bind \texttt{aSnpMatrix} objects using the S4 \texttt{cbind2} and
\texttt{rbind2} methods:

<<>>=
i2 <- 21:40
j2 <- 101:200
as12 <- new("aSnpMatrix",
                 .Data=Autosomes[i1,j2],
                 snps=Asnps[j2,,drop=FALSE],
                 samples=subject.data[i1,])
as21 <- new("aSnpMatrix",
                 .Data=Autosomes[i2,j1],
                 snps=Asnps[j1,,drop=FALSE],
                 samples=subject.data[i2,])

cbind2(as11,as12)
rbind2(as11,as21)
@ %def
\section{Other functions}
\label{sec-3}

A few other handy functions are included. See the individual help
pages for more detail on these.

\begin{itemize}
\item \texttt{dups} finds indices of possible sample duplications between two
aSnpStats objects
\item \texttt{mismatch.count} quickly counts mismatched genotypes between pairs
of individuals
\item \texttt{sample.match} identifies indices of overlapping samples in two
aSnpMatrix objects.  By default, match is on \texttt{rownames(x@samples)},
\texttt{rownames(y@samples)}.
\item \texttt{snp.match} identifies indices of overlapping samples in two
aSnpMatrix objects.  By default, match is on \texttt{rownames(x@snps)},
\texttt{rownames(y@snps)}.
\item \texttt{snp.trim} drops SNPs from an annotSnpStats object according to SNP qc summary
stats
\end{itemize}
% Emacs 24.2.1 (Org mode 8.0.3)
\end{document}
