#+TITLE:     annotSnpStats: a package for annotated SnpMatrix objects
#+AUTHOR:    Chris Wallace
#+EMAIL:     chris.wallace@cimr.cam.ac.uk
#+DATE:      \today
#+DESCRIPTION:
#+KEYWORDS:
#+LANGUAGE:  en
#+OPTIONS:   H:3 num:t toc:t \n:nil @:t ::t |:t ^:t -:t f:t *:t <:t
#+OPTIONS:   TeX:t LaTeX:t skip:nil d:(not LOGBOOK) todo:t pri:nil tags:t

#+EXPORT_SELECT_TAGS: export
#+EXPORT_EXCLUDE_TAGS: noexport
#+LINK_UP:
#+LINK_HOME:
#+XSLT:

#+latex_header: \usepackage{fullpage}
#+latex: %\VignetteIndexEntry{annotated SnpMatrix objects}

* Introduction

=annotSnpStats= defines the =aSnpMatrix= and =aXSnpMatrix= classes
which extend the =SnpMatrix= and =XSnpMatrix= classes from =snpStats=
by aligning genotype data with two dataframes containing SNP (column)
specific and sample (row) specific information.  Subsetting, binding
etc can then be carried out keeping this additional information
aligned.

If you don't know what the =snpStats= package does, then this package
is unlikely to be of use.  But if you use =snpStats=, then, hopefully,
annotSnpStats will make lots of things easier.  Various utility
functions have been added which aim to make the manipulation and
merging of two or more genotype datasets as painless and as accurate
as possible.

* Basic data structure

aSnpMatrix objects consist, at a minimum, of a data matrix
(=(X)SnpMatrix=), and row-wise sample information =data.frame= and and
column-wise SNP information =data.frame=.  The =example.data()=
function can help you create test objects, but looking at the code
inside the function can help you create objects from your own data
too.

#+begin_src R :ravel 
library(annotSnpStats)
## see how an object can be created
example.data()

## create one
X <- example.data(20,100)
X
#+end_src

Optionally, you can set a =phenotype()= which records which column of
the samples =data.frame= contains a column describing a sample's
phenotype.  =example.data()= already does this, but you could
change/set it thus:

* Phenotype

#+begin_src R :ravel 
phenotype(X) <- "cc"
#+end_src

* Allele codes
Genotype data is stored in a (compressed) format with 1=AA, 2=AB and
3=BB genotypes.  If you wish to merge two datasets, you need to know
what each code corresponds to.  This information should be stored in a
pair of columns in the =snps= slot of the object.  The first member
should show the base (A, C, G or T) corresponding to an AA homozygote,
and the second the base corresponding to a BB homozygote.  =aSnpStats=
functions can try and guess which columns might be storing this
information, but it is safer to specify explicitly.

#+begin_src R :ravel 
# set the allele codes
alleles(X) <- c("A1","A2")
#+end_src

* Examples of use

The advantage of using =annotSnpStats= is perhaps best seen using some
example data from the =snpStats= package to define some new
=aSnpMatrix= objects and play with them.

** Access annotation objects

#+begin_src R :ravel 
head(snps(X))
head(samples(X))
#+end_src

** Use in place of regular SnpMatrix objects

All =snpStats= S4 methods should work with =aSnpMatrix= objects

#+begin_src R :ravel 
summary(X)
#+end_src

However, certain =snpStats= functions do not use the S4 interface, and
inheritance then gets a bit messed up.  In that case, you may need to
explicity extract the =SnpMatrix= object.  This can be done with =as(,Class="SnpMatrix")=
or, for convenience, use =sm()=
For example, to use
snpStats's =ld= function:

#+begin_src R :ravel 
ld(sm(X)[1:5,1:5],depth=3,stats="R.squared")
#+end_src

** Subsetting

Subsetting subsets both the =SnpMatrix= and the annotation objects

#+begin_src R :ravel 
asub <- X[1:3,1:5]
dim(X)
dim(asub)
nrow(snps(X))
nrow(snps(asub))
nrow(samples(X))
nrow(samples(asub))
#+end_src

** Binding
You can c/rbind =aSnpMatrix= objects using the S4 =cbind2= and
=rbind2= methods:

#+begin_src R :ravel 
## add additional SNPs
Xw <- example.data(20,101:200)
cbind2(X,Xw)

## add additional samples
Xl <- example.data(21:40,100)
rbind2(X,Xl)

#+end_src

* Matching SNPs and samples between datasets

annotSnpStats was created primarily to make the manipulation and
joining of multiple genetic datasets easier. Identifying which samples
and SNPs are duplicated between datasets can be very important to
avoid double counting.

=sample.match= identifies indices of overlapping samples in two
  aSnpMatrix objects.  By default, match is on =rownames(samples(x))=,
  =rownames(samples(y))=.

#+BEGIN_SRC R :ravel
sample.match(X,Xw) # should have matches
sample.match(X,Xl) # should have no matches
#+END_SRC

=snp.match= identifies indices of overlapping samples in two
  aSnpMatrix objects.  By default, match is on =rownames(snps(x))=,
  =rownames(snps(y))=.

#+BEGIN_SRC R :ravel
snp.match(X,Xw) # should have no matches
snp.match(X,Xl) # should have matches
#+END_SRC

However, often the same samples can appear in different datasets with
different ids. =dups= finds indices of possible sample duplications
between two aSnpStats objects by comparing genotypes. By default, two
homozygote genotypes are called a match, regardless of which
homozygotes they are.  This is useful if you are not sure that allele
codes are comparable across the datasets.

#+BEGIN_SRC R :ravel
Y <- example.data(30:11,100)
dups(X,Y) # should match rows 11:20 of X with rows 20:11 of Y

## now, reorder the allele labelling
Y <- switch.alleles(Y,1:50)
dups(X,Y) # should STILL match rows 11:20 of X with rows 1:10 of Y
#+END_SRC

You can specify that matches take account of all genotype states, not just hethom (the default) by

#+BEGIN_SRC R :ravel
dups(X,Y,type="all") # should find NO matches
#+END_SRC

* Aligning alleles

Differences in allele coding is one of the most frustrating problems
with dealing with genetic data from different sources.  One issue is
simple differences in numeric coding, eg in =X= allele A is 1 and
allele G is 2, whereas in =Y= allele A is 2 and G is 1.  I call this a
code reverse.  However, sometimes alleles can be recorded from
different strands, so in =Y= allele T is 1 and allele C is 2, this is
a "complement".  Just for fun, we sometimes have both happen.

#+BEGIN_SRC R :ravel
alleles.X <- c(snp1="A/G",snp2="A/G",snp3="A/G",snp4="A/G")
alleles.Y <- c(snp1="A/G",snp2="G/A",snp3="T/C",snp4="C/T")
classes <- g.class(x=alleles.X,y=alleles.Y)
cbind(alleles.X,alleles.Y,classes)
#+END_SRC

This is ok when a SNP is A/G, each class can be resolved.  But an
A/T SNP is harder, because a reverse looks the same as a complement:

#+BEGIN_SRC R :ravel
alleles.X <- c(snp1="A/T",snp2="A/T",snp3="A/T",snp4="A/T")
alleles.Y <- c(snp1="A/T",snp2="T/A",snp3="T/A",snp4="A/T")
classes <- g.class(x=alleles.X,y=alleles.Y)
cbind(alleles.X,alleles.Y,classes)
#+END_SRC

annotSnpStats first uses information from SNPs which have unambiguous
allele codes.  If there are no strand switches amongst these SNPs, it
assumes everything is on the same strand and resolveable.  A
diagnostic plot is useful to check the alignment has worked.

#+BEGIN_SRC R :ravel fig=TRUE
Y.aligned <- align.alleles(X,Y)
#+END_SRC

Where there are strand switches, annotSnpStats uses allele frequencies
to try and resolve the ambigous alignments, but this can only work
where (a) sample sizes are reasonably large and from the same
populations and (b) minor allele frequencies not too close to 0.5.
Some inference can also be made if you have know duplicates between
the two datasets, thus:

#+BEGIN_SRC R :ravel fig=TRUE
XY.dups <- dups(X,Y)
if(nrow(XY.dups)) {
  Y.aligned <- align.alleles(X,Y,known.dups=XY.dups)
} else {
  Y.aligned <- align.alleles(X,Y)
}
#+END_SRC


* Other functions

=snp.trim= drops SNPs from an annotSnpStats object according to SNP
qc summary stats according to user specified limits.  See help file.
 
# - =mismatch.count= quickly counts mismatched genotypes between pairs
#   of individuals
# - =mismatch.count.all= quickly counts mismatched genotypes between all
#   possible pairs of individuals - this can be slow!
